lib/utils/interval/multi.go

```diff
@@ -0,0 +1,397 @@
+	d := s.Duration
+	if s.Jitter != nil {
+		d = s.Jitter(d)
+	}
+	s.next = now.Add(d)
+}
+
+func (s *subIntervalEntry[T]) increment() {
+	s.reset(s.next)
+}
+
+// NewMulti creates a new multi-interval instance.  This function panics on non-positive
+// interval durations (equivalent to time.NewTicker) or if no sub-intervals are provided.
+func NewMulti[T comparable](intervals ...SubInterval[T]) *MultiInterval[T] {
+	if len(intervals) == 0 {
+		panic(errors.New("empty sub-interval set for interval.NewMulti"))
+	}
+
+	interval := &MultiInterval[T]{
+		subs:  make([]subIntervalEntry[T], 0, len(intervals)),
+		push:  make(chan subIntervalEntry[T]),
+		ch:    make(chan Tick[T], 1),
+		reset: make(chan T),
+		fire:  make(chan T),
+		done:  make(chan struct{}),
+	}
+
+	// check and initialize our sub-intervals.
+	now := time.Now()
+	for _, sub := range intervals {
+		if sub.Duration <= 0 {
+			panic(errors.New("non-positive sub interval for interval.NewMulti"))
+		}
+		entry := subIntervalEntry[T]{
+			SubInterval: sub,
+		}
+		entry.init(now)
+		interval.pushEntry(entry)
+	}
+
+	key, d := interval.duration(now)
+
+	// start the timer in this goroutine to improve
+	// consistency of first tick.
+	timer := time.NewTimer(d)
+
+	go interval.run(timer, key)
+
+	return interval
+}
+
+// Push adds a new sub-interval, potentially overwriting an existing sub-interval with the same key.
+// This method panics on non-positive durations (equivalent to time.NewTicker).
+func (i *MultiInterval[T]) Push(sub SubInterval[T]) {
+	if sub.Duration <= 0 {
+		panic(errors.New("non-positive sub interval for MultiInterval.Push"))
+	}
+	entry := subIntervalEntry[T]{
+		SubInterval: sub,
+	}
+	// we initialize here in order to improve consistency of start time
+	entry.init(time.Now())
+	select {
+	case i.push <- entry:
+	case <-i.done:
+	}
+}
+
+// Stop permanently stops the interval.  Note that stopping an interval does not
+// close its output channel.  This is done in order to prevent concurrent stops
+// from generating erroneous "ticks" and is consistent with the behavior of
+// time.Ticker.
+func (i *MultiInterval[T]) Stop() {
+	i.closeOnce.Do(func() {
+		close(i.done)
+	})
+}
+
+// Reset resets the interval without pausing it (i.e. it will now fire in
+// jitter(duration) regardless of current timer progress).
+func (i *MultiInterval[T]) Reset(key T) {
+	select {
+	case i.reset <- key:
+	case <-i.done:
+	}
+}
+
+// FireNow forces the sub-interval to fire immediately regardless of how much time is left on
+// the current interval. This also effectively resets the sub-interval.
+func (i *MultiInterval[T]) FireNow(key T) {
+	select {
+	case i.fire <- key:
+	case <-i.done:
+	}
+}
+
+// Next is the channel over which interval ticks are delivered.
+func (i *MultiInterval[T]) Next() <-chan Tick[T] {
+	return i.ch
+}
+
+// duration gets the next duration and its associated key.
+func (i *MultiInterval[T]) duration(now time.Time) (key T, d time.Duration) {
```

Comment:
How many subintervals do we expect? This is going to be linear in the amount of subintervals, but we could also maintain a binary heap keyed by the next tick in logarithmic time for all operations.

Reply:
Right now... I can't force any use-case that would incur more sub-intervals than N+1 where N is the number of teleport services.  Binary heap is a good idea, but doubt it's worth the effort unless we start using this in a context that demands hundreds of sub-intervals.  Maybe more.